src/ext/java/org/opentripplanner/ext/flex/FlexRouter.java

```diff
@@ -0,0 +1,200 @@
+import java.time.LocalDate;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class FlexRouter {
+
+  /* Transit data */
+  private final Graph graph;
+  private final Collection<StopAtDistance> streetAccesses;
+  private final Collection<StopAtDistance> streetEgresses;
+  private final StopIndexForRaptor stopIndex;
+  private final Collection<Transfer> transitTransfers;
+  private final FlexIndex flexIndex;
+  private final DistanceCalculator distanceCalculator;
+
+  /* Request data */
+  private final ZonedDateTime startOfTime;
+  private final int departureTime;
+  private final boolean arriveBy;
+
+  private final TIntSet[] servicesRunning;
+  private final int[] differenceFromStartOfTime;
+  private final ServiceDate[] serviceDates;
+
+  /* State */
+  private List<FlexAccessTemplate> flexAccessTemplates = null;
+  private List<FlexEgressTemplate> flexEgressTemplates = null;
+
+  public FlexRouter(
+      RoutingRequest request,
+      int additionalPastSearchDays,
+      int additionalFutureSearchDays,
+      Collection<StopAtDistance> streetAccesses,
+      Collection<StopAtDistance> egressTransfers,
+      StopIndexForRaptor stopIndex
+  ) {
+    this.graph = request.rctx.graph;
+    this.streetAccesses = streetAccesses;
+    this.streetEgresses = egressTransfers;
+    this.stopIndex = stopIndex;
+    this.transitTransfers = graph.getTransferTable().getTransfers();
+    this.flexIndex = graph.index.getFlexIndex();
+    this.distanceCalculator = new DirectDistanceCalculator(graph);
+
+    ZoneId tz = graph.getTimeZone().toZoneId();
+    Instant searchInstant = request.getDateTime().toInstant();
+    LocalDate searchDate = LocalDate.ofInstant(searchInstant, tz);
+    this.startOfTime = DateMapper.asStartOfService(searchDate, tz);
+    this.departureTime = DateMapper.secondsSinceStartOfTime(startOfTime, searchInstant);
+    this.arriveBy = request.arriveBy;
+
+    int totalDays = additionalPastSearchDays + 1 + additionalFutureSearchDays;
+
+    this.differenceFromStartOfTime = new int[totalDays];
+    this.servicesRunning = new TIntSet[totalDays];
+    this.serviceDates = new ServiceDate[totalDays];
+
+    for (int d = -additionalPastSearchDays; d <= additionalFutureSearchDays; ++d) {
+      LocalDate date = searchDate.plusDays(d);
+      int index = d + additionalPastSearchDays;
+      ServiceDate serviceDate = new ServiceDate(date);
+      differenceFromStartOfTime[index] = DateMapper.secondsSinceStartOfTime(startOfTime, date);
+      servicesRunning[index] = graph.index.getServiceCodesRunningForDate().get(serviceDate);
+      serviceDates[index] = serviceDate;
+    }
+  }
+
+  private void calculateFlexAccessTemplates() {
```

Comment:
Maybe add some JavaDoc on this method to explain the the intention of the method and any optimalizations.